# PyCity Schools Analysis

* As a whole, schools with higher budgets, did not yield better test results. By contrast, schools with higher spending per student actually (\$645-675) underperformed compared to schools with smaller budgets (<\$585 per student).

* As a whole, smaller and medium sized schools dramatically out-performed large sized schools on passing math performances (89-91% passing vs 67%).

* As a whole, charter schools out-performed the public district schools across all metrics. However, more analysis will be required to glean if the effect is due to school practices or the fact that charter schools tend to serve smaller student populations per school. 
---

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [356]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas Data Frames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the overall passing rate (overall average score), i.e. (avg. math score + avg. reading score)/2

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting



In [357]:
# Calculate the total number of schools
tot_school = len(school_data_complete['school_name'].unique())
tot_school

15

Total number of schools is 15.

In [358]:
# Calculate the total number of students
tot_student = len(school_data_complete['Student ID'].unique())
tot_student

39170

The total number of Students is 39170.

In [359]:
# Calculate the total budget
tot_budget = school_data['budget'].sum()
tot_budget

24649428

The total budget is 82932329558.

In [360]:
# Calculate the average math score
avg_math = round( school_data_complete['math_score'].mean(), 2)
avg_math

78.99

The average math score is 78.99.

In [361]:
# Calculate the average reading score
avg_read = round(school_data_complete['reading_score'].mean(), 2)
avg_read

81.88

The average reading score is 81.88.

In [362]:
# Calculate the overall passing rate (overall average score), i.e. (avg. math score + avg. reading score)/2
pass_overall = round((school_data_complete['math_score'].mean()+school_data_complete['reading_score'].mean())/2, 2)
pass_overall

80.43

The overall passing rate is 80.43.

In [363]:
# Calculate the percentage of students with a passing math score (70 or greater)
pass_math = len(school_data_complete[school_data_complete.math_score>=70]['math_score'])/len(school_data_complete.math_score)*100
pass_math = round(pass_math, 2)
pass_math

74.98

The percentage of students with a passing math score is 74.98.

In [364]:
# Calculate the percentage of students with a passing reading score (70 or greater)
pass_read = len(school_data_complete[school_data_complete.reading_score>=70]['reading_score'])/len(school_data_complete.math_score)*100
pass_read = round(pass_read, 2)
pass_read

85.81

The percentage of students with a passing reading score is 85.81.

In [365]:
# Create a dataframe to hold the above results
df_result = pd.DataFrame({"Total Schools":[tot_school],"Total Students":[tot_student],
                          "Total Budget":[tot_budget],"Average Math Score": [avg_math],"Average Reading Score":[avg_read],
                          "% Passing Math":[pass_math],"% Passing Reading" :[pass_read],"% Overall Passing Rate":[pass_overall]})


In [366]:
# Optional: give the displayed data cleaner formatting
strr = '${:,}'.format(df_result['Total Budget'].astype(int)[0])
print(strr)
df_result['Total Budget'] = str(strr)
df_result

$24,649,428


Total Schools  Total Students Total Budget  Average Math Score  \
0             15           39170  $24,649,428               78.99   

   Average Reading Score  % Passing Math  % Passing Reading  \
0                  81.88           74.98              85.81   

   % Overall Passing Rate  
0                   80.43

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)
  
* Create a dataframe to hold the above results

In [367]:
# * Create an overview table that summarizes key metrics about each school, including:
#   * School Name
#   * School Type
#   * Total Students
tot_student = school_data_complete[['school_name','size']].groupby(['school_name']).count()
tot_student = tot_student.rename(columns = {'size':'Total Students'})
tot_student = tot_student[['Total Students']]

In [368]:
#   * Total School Budget
tot_budget = school_data[['school_name','budget']].groupby(['school_name']).sum()
tot_budget = tot_budget.rename(columns = {'budget':'Total Budget'})
tot_budget = tot_budget[['Total Budget']]

In [369]:
#   * Per Student Budget
per_student = tot_budget['Total Budget']/ tot_student['Total Students']
per_student = pd.DataFrame(per_student, columns=["Per Student Budget"])

In [370]:
#   * Average Reading Score
#   * Average Math Score
avg_math_read = school_data_complete[['math_score','reading_score','school_name']].groupby(['school_name']).mean()
avg_math_read.rename(columns={"math_score":"Average Math Score","reading_score":"Average Reading Score"}, inplace=True)

In [371]:
#   * % Passing Math
school_data_pass = school_data_complete[school_data_complete.math_score>= 70]
math_pass = school_data_pass[['school_name','math_score']].groupby(['school_name']).count()
math_total = school_data_complete[['school_name','math_score']].groupby(['school_name']).count()
percent_math_pass = math_pass['math_score']/ math_total['math_score']*100
percent_math_pass = pd.DataFrame(percent_math_pass)
percent_math_pass.rename(columns={"math_score":"% Passing Math"}, inplace=True)

In [372]:
#   * % Passing Reading
school_data_pass = school_data_complete[school_data_complete.reading_score>= 70]
read_pass = school_data_pass[['school_name','reading_score']].groupby(['school_name']).count()
read_total = school_data_complete[['school_name','reading_score']].groupby(['school_name']).count()
percent_read_pass = read_pass['reading_score']/ read_total['reading_score']*100
percent_read_pass = pd.DataFrame(percent_read_pass)
percent_read_pass.rename(columns={"reading_score":"% Passing Reading"}, inplace=True)

In [373]:
#   * Overall Passing Rate (Average of the above two)
overall_pass_rate = (percent_math_pass[['% Passing Math',]].values + percent_read_pass[['% Passing Reading']].values)/2
overall_pass_rate = pd.DataFrame(overall_pass_rate)
overall_pass_rate.rename(columns={0:"% Overall Passing Rate"}, inplace=True)
overall_pass_rate.index = percent_math_pass.index
overall_pass_rate

% Overall Passing Rate
school_name                                  
Bailey High School                  74.306672
Cabrera High School                 95.586652
Figueroa High School                73.363852
Ford High School                    73.804308
Griffin High School                 95.265668
Hernandez High School               73.807983
Holden High School                  94.379391
Huang High School                   73.500171
Johnson High School                 73.639992
Pena High School                    95.270270
Rodriguez High School               73.293323
Shelton High School                 94.860875
Thomas High School                  95.290520
Wilson High School                  95.203679
Wright High School                  94.972222

In [374]:
school_type = school_data[['school_name','type']]
school_type.reset_index(inplace=True)

In [375]:
  
# * Create a dataframe to hold the above results
results = pd.concat(
[tot_student, 
tot_budget,
per_student,
avg_math_read,
percent_math_pass,
percent_read_pass,
overall_pass_rate
],
axis = 1)

In [376]:
results.head()

Total Students  Total Budget  Per Student Budget  \
school_name                                                              
Bailey High School              4976       3124928               628.0   
Cabrera High School             1858       1081356               582.0   
Figueroa High School            2949       1884411               639.0   
Ford High School                2739       1763916               644.0   
Griffin High School             1468        917500               625.0   

                      Average Math Score  Average Reading Score  \
school_name                                                       
Bailey High School             77.048432              81.033963   
Cabrera High School            83.061895              83.975780   
Figueroa High School           76.711767              81.158020   
Ford High School               77.102592              80.746258   
Griffin High School            83.351499              83.816757   

                      % Passing Math  % Passing Reading  \
school_name                                               
Bailey High School         66.680064          81.933280   
Cabrera High School        94.133477          97.039828   
Figueroa High School       65.988471          80.739234   
Ford High School           68.309602          79.299014   
Griffin High School        93.392371          97.138965   

                      % Overall Passing Rate  
school_name                                   
Bailey High School                 74.306672  
Cabrera High School                95.586652  
Figueroa High School               73.363852  
Ford High School                   73.804308  
Griffin High School                95.265668

In [377]:
school_type.reset_index(inplace=True)
results.reset_index(inplace=True)
results.index = results.school_name
school_type.index = school_type.school_name


In [378]:
results = pd.concat([school_type, results], axis = 1)

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=True'.

To retain the current behavior and silence the warning, pass sort=False

  """Entry point for launching an IPython kernel.


In [379]:
results.columns

Index(['level_0', 'index', 'school_name', 'type', 'school_name',
       'Total Students', 'Total Budget', 'Per Student Budget',
       'Average Math Score', 'Average Reading Score', '% Passing Math',
       '% Passing Reading', '% Overall Passing Rate'],
      dtype='object')

In [380]:
results = results[[ 'type',
       'Total Students', 'Total Budget', 'Per Student Budget',
       'Average Math Score', 'Average Reading Score',
       '% Passing Reading', '% Passing Math','% Overall Passing Rate']]
results.head()

type  Total Students  Total Budget  \
Bailey High School    District            4976       3124928   
Cabrera High School    Charter            1858       1081356   
Figueroa High School  District            2949       1884411   
Ford High School      District            2739       1763916   
Griffin High School    Charter            1468        917500   

                      Per Student Budget  Average Math Score  \
Bailey High School                 628.0           77.048432   
Cabrera High School                582.0           83.061895   
Figueroa High School               639.0           76.711767   
Ford High School                   644.0           77.102592   
Griffin High School                625.0           83.351499   

                      Average Reading Score  % Passing Reading  \
Bailey High School                81.033963          81.933280   
Cabrera High School               83.975780          97.039828   
Figueroa High School              81.158020          80.739234   
Ford High School                  80.746258          79.299014   
Griffin High School               83.816757          97.138965   

                      % Passing Math  % Overall Passing Rate  
Bailey High School         66.680064               74.306672  
Cabrera High School        94.133477               95.586652  
Figueroa High School       65.988471               73.363852  
Ford High School           68.309602               73.804308  
Griffin High School        93.392371               95.265668

## Top Performing Schools (By Passing Rate)

* Sort and display the top five schools in overall passing rate

In [381]:
results.sort_values(['% Overall Passing Rate'], ascending=False).head()

type  Total Students  Total Budget  \
Cabrera High School  Charter            1858       1081356   
Thomas High School   Charter            1635       1043130   
Pena High School     Charter             962        585858   
Griffin High School  Charter            1468        917500   
Wilson High School   Charter            2283       1319574   

                     Per Student Budget  Average Math Score  \
Cabrera High School               582.0           83.061895   
Thomas High School                638.0           83.418349   
Pena High School                  609.0           83.839917   
Griffin High School               625.0           83.351499   
Wilson High School                578.0           83.274201   

                     Average Reading Score  % Passing Reading  % Passing Math  \
Cabrera High School              83.975780          97.039828       94.133477   
Thomas High School               83.848930          97.308869       93.272171   
Pena High School                 84.044699          95.945946       94.594595   
Griffin High School              83.816757          97.138965       93.392371   
Wilson High School               83.989488          96.539641       93.867718   

                     % Overall Passing Rate  
Cabrera High School               95.586652  
Thomas High School                95.290520  
Pena High School                  95.270270  
Griffin High School               95.265668  
Wilson High School                95.203679

## Bottom Performing Schools (By Passing Rate)

* Sort and display the five worst-performing schools

In [382]:
results.sort_values(['% Overall Passing Rate']).head()

type  Total Students  Total Budget  \
Rodriguez High School  District            3999       2547363   
Figueroa High School   District            2949       1884411   
Huang High School      District            2917       1910635   
Johnson High School    District            4761       3094650   
Ford High School       District            2739       1763916   

                       Per Student Budget  Average Math Score  \
Rodriguez High School               637.0           76.842711   
Figueroa High School                639.0           76.711767   
Huang High School                   655.0           76.629414   
Johnson High School                 650.0           77.072464   
Ford High School                    644.0           77.102592   

                       Average Reading Score  % Passing Reading  \
Rodriguez High School              80.744686          80.220055   
Figueroa High School               81.158020          80.739234   
Huang High School                  81.182722          81.316421   
Johnson High School                80.966394          81.222432   
Ford High School                   80.746258          79.299014   

                       % Passing Math  % Overall Passing Rate  
Rodriguez High School       66.366592               73.293323  
Figueroa High School        65.988471               73.363852  
Huang High School           65.683922               73.500171  
Johnson High School         66.057551               73.639992  
Ford High School            68.309602               73.804308

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [383]:
math_grade = school_data_complete[['school_name', 'math_score','grade']].groupby(['school_name','grade']).mean()
math_grade.reset_index(inplace= True)
math_grade = math_grade.pivot('school_name','grade','math_score')
math_grade = math_grade[['9th','10th','11th','12th']]
math_grade

grade                        9th       10th       11th       12th
school_name                                                      
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [384]:
read_grade = school_data_complete[['school_name', 'reading_score','grade']].groupby(['school_name','grade']).mean()
read_grade.reset_index(inplace= True)
read_grade = read_grade.pivot('school_name','grade','reading_score')
read_grade = read_grade[['9th','10th','11th','12th']]
read_grade

grade                        9th       10th       11th       12th
school_name                                                      
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [ ]:
school_data_spending = results

In [421]:
# Sample bins. Feel free to create your own bins.
spending_bins = [0, 590, 620, 650, 675]
group_names = ["<$590", "$590-620", "$620-650", "$650-675"]

In [422]:
school_data_spending['spending_group'] = pd.cut(school_data_spending['Per Student Budget'], bins=spending_bins, labels= group_names)
score_spending = school_data_spending.groupby(['spending_group']).mean()
score_spending = score_spending[['Average Math Score', 'Average Reading Score', '% Passing Reading',
       '% Passing Math', '% Overall Passing Rate']]
score_spending

Average Math Score  Average Reading Score  % Passing Reading  \
spending_group                                                                 
<$590                    83.455399              83.933814          96.610877   
$590-620                 83.599686              83.885211          95.900287   
$620-650                 78.792545              81.759287          85.408835   
$650-675                 76.959583              81.058567          81.089710   

                % Passing Math  % Overall Passing Rate  
spending_group                                          
<$590                93.460096               95.035486  
$590-620             94.230858               95.065572  
$620-650             74.295260               79.852048  
$650-675             66.218444               73.654077

## Scores by School Size

* Perform the same operations as above, based on school size.

In [476]:
# Sample bins. Feel free to create your own bins.
size_bins = [0, 1200, 2400, 5000]
group_names = ["Small (<1200)", "Medium (1200-2400)", "Large (2400-5000)"]

scores_data = pd.merge(results, school_data)

scores_data['size_group'] = pd.cut(scores_data['size'], bins=size_bins, labels= group_names)


scores_data = scores_data.groupby(['size_group']).mean()
scores_data = scores_data[['Average Math Score', 'Average Reading Score', '% Passing Reading',
       '% Passing Math', '% Overall Passing Rate']]
scores_data

Average Math Score  Average Reading Score  \
size_group                                                      
Small (<1200)                83.473852              83.896421   
Medium (1200-2400)           83.473852              83.896421   
Large (2400-5000)            76.956733              80.966636   

                    % Passing Reading  % Passing Math  % Overall Passing Rate  
size_group                                                                     
Small (<1200)               96.586489       93.620830               95.103660  
Medium (1200-2400)          96.586489       93.620830               95.103660  
Large (2400-5000)           80.799062       66.548453               73.673757

## Scores by School Type

* Perform the same operations as above, based on school type.

In [481]:
type_data = pd.merge(results, school_data)

In [484]:
# Sample bins. Feel free to create your own bins.
type_data = pd.merge(results, school_data)
type_data = type_data.groupby(['type']).mean()
type_data = type_data[['Average Math Score', 'Average Reading Score', '% Passing Reading',
       '% Passing Math', '% Overall Passing Rate']]
type_data

Average Math Score  Average Reading Score  % Passing Reading  \
type                                                                     
Charter            83.473852              83.896421          96.586489   
District           76.956733              80.966636          80.799062   

          % Passing Math  % Overall Passing Rate  
type                                              
Charter        93.620830               95.103660  
District       66.548453               73.673757

#  Written description of at least two observable trends based on the data

Observable trends are
* Top performing 5 schools are charter but bottom performing 5 schools are district. 
* From math score and reading score by grade we can see that students permoed lower in math score in some cases and in some cases higher. So math score varies much.
But for reading score all student got on average above 80. So some student's struggle with math but all are comforatable with reading.
* From Scores by School Size we can see that performance of schools decreases with incresing size. The following scores are lower 
with the higher school size.
Average Math Score,Average Reading Score,% Passing Reading,% Passing Math,% Overall Passing Rate 
* From Scores by School Type data average scores we can see that charter schools student's scored higher than district school student on average.